core/src/test/java/org/infinispan/persistence/dummy/DummyInMemoryStore.java

```diff
@@ -281,22 +276,19 @@ private boolean isExpired(MarshallableEntry me, long now) {
    public Flowable<MarshallableEntry> publishEntries(IntSet segments, Predicate filter, boolean fetchValue) {
       assertRunning();
       record("publishEntries");
-      log.tracef("Publishing entries in store %s with filter %s", storeName, filter);
+      log.tracef("Publishing entries in store %s segments %s with filter %s", storeName, segments, filter);
       Flowable<Map.Entry<Object, byte[]>> flowable;
       if (configuration.segmented()) {
-         flowable = Flowable.fromStream(segments.intStream().mapToObj(segment -> {
-            Map<Object, byte[]> map = store.get(segment);
-            if (map == null) {
-               return Flowable.<Map.Entry<Object, byte[]>>empty();
-            }
-            return Flowable.fromIterable(map.entrySet());
-         })).flatMap(RxJavaInterop.identityFunction());
+         flowable = Flowable.fromIterable(segments)
+                 .concatMap(segment -> {
+                    Map<Object, byte[]> map = store.get(segment);
+                    if (map == null) {
+                       return Flowable.<Map.Entry<Object, byte[]>>empty();
+                    }
+                    return Flowable.fromIterable(map.entrySet());
+                 });
       } else {
-         flowable = Flowable.fromIterable(store.get(0).entrySet())
-               .filter(e -> {
-                  int segment = keyPartitioner.getSegment(e.getKey());
-                  return segments.contains(segment);
-               });
+         flowable = Flowable.fromIterable(store.get(0).entrySet());
```

Comment:
Why was the filter removed? It should be more efficient to filter by segment here and not deserialize then to deserialize every entry and then have it filtered at PersistenceManagerImpl.

Reply:
My thinking was that the store is not going to be segmented when the user doesn't need segments, so the publish methods are going to be invoked mostly with the full set of segments and the segment check is going to be just extra work.

Reply:
Just because a store isn't segmented doesn't mean the user doesn't need segments. In many cases this is called with a subset of segments when it is not shared due to topology changes.

But looking closer this change is fine. I forgot that when the store is not segmented we filter out the segments of the keys in the provided filter, so the lines below will handle this anyways. https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L740

Reply:
Yeah, I should have been clearer when I said "extra work", I meant "duplicated work" because PMI assumes the store doesn't filter by segment.

And by "the user doesn't need segments" I meant that the user must have set segmented="false" because they know they don't need segments. If they do need segments and they still set segmented="false", then the unmarshalling of the value is just a drop in the bucket compared to all the iteration they have to perform over all their segments, when publishers request only 1 segment at a time.

Reply:
> Yeah, I should have been clearer when I said "extra work", I meant "duplicated work" because PMI assumes the store doesn't filter by segment.

It actually does with the provided filter as I mentioned in my previous comment :)

> And by "the user doesn't need segments" I meant that the user must have set segmented="false" because they know they don't need segments. If they do need segments and they still set segmented="false", then the unmarshalling of the value is just a drop in the bucket compared to all the iteration they have to perform over all their segments, when publishers request only 1 segment at a time.

It is definitely _not_ a drop in the bucket. You are talking about deserializing every entry for every segment, no matter what versus only the ones you care about. Iterating and deserializing N entries versus iterating N and deserializing N/2 is a BIG difference (assuming only half the segments are requested).

But as I stated in my previous comment this change is fine as we already filter out the entries in the provided filter automatically for non segmented stores.

Reply:
> You are talking about deserializing every entry for every segment, no matter what versus only the ones you care about.

Sorry I wasn't clear again, but I did not mean that: I was assuming that PMI filters by segment before unmarshalling, if the store is not segmented (because sometimes I trusted the javadoc and sometimes I looked at the PMI code instead, so I wasn't very consistent).

The point I was trying to get to is that a user who cares about performance in a clustered cache will not set segmented="false". If they do disable segmentation for "reasons", my optimization of hashing each key only in the PMI instead of once in the PMI and once in the store won't make a difference in their choice.

Reply:
Sounds like we are on the same page then :)

But there is no reason to has in the store as well for this case as the PMI handles the case and it is less code, so removing it 👍